pkg/codegen/pcl/invoke.go

```diff
@@ -44,6 +44,56 @@ func getInvokeToken(call *hclsyntax.FunctionCallExpr) (string, hcl.Range, bool)
 	return literal.Val.AsString(), call.Args[0].Range(), true
 }
 
+// annotateObjectProperties annotates the properties of an object expression with the types of the corresponding
+// properties in the schema. This is used to provide type information
+// for invoke calls that didn't have type annotations.
+//
+// this function will recursively annotate the properties of objects
+// that are nested within the object expression type.
+func annotateObjectProperties(modelType model.Type, schemaType schema.Type, topLevelAnnotate bool) {
+	if optionalType, ok := schemaType.(*schema.OptionalType); ok {
+		schemaType = optionalType.ElementType
+	}
+
+	if objectType, ok := modelType.(*model.ObjectType); ok {
```

Comment:
Since we don't expect more than one of these `if _, ok := modelType.(_); ok` to trigger, a switch statement would be cleaner.
```suggestion
switch modelType := modelType.(type) {
case *model.ObjectType:
```